### DISCUSSION TITLE
Portable build.

### DISCUSSION THREAD

**Discussion by ghost** [input-leap/input-leap#1734] [Ideas]
> This issue has been migrated from old Barrier Github repository https://github.com/debauchee/barrier/issues/961
> 
> Issue created on: 2020-11-28 by @Metroid-Prime
> Issue last updated on: 2020-12-05

Hello,
I've noticed Barrier stores its settings in multiple folders. This is my #1 request for this beautiful & useful software.

Settings are stored in: 
C:\Program Files\Barrier
C:\Users\user\AppData\Local\Barrier
C:\ProgramData\Barrier

Is it possible implement the -portable command argument so that I can run Barrier from another directory and all settings get stored in the program directory? **Pretty please?**
D:\Applications\Barrier

---

> Commented on: 2020-12-01 by @ArmandoRl1

This would be wonderful to have

---

> Commented on: 2020-12-04 by @albertony

I agree, an easy to use, official portable mode would be nice! Combined with a zip distribution for copy-deploy.

If you can manage with just the command line editions (barriers.exe and barrierc.exe), I think it is possible already:
- When running the client and server, specify path to where you want your settings stored with command line argument `--profile-dir <path>`.
   - In your case `--profile-dir D:\Applications\Barrier`.
   - For example: `D:\Applications\Barrier\barriers.exe --profile-dir D:\Applications\Barrier --no-daemon --no-tray --enable-crypto`
   - The default path, when this argument is not specified, will otherwise be `%LocalAppdata%\Barrier`.
- On the server, if you name the server configuration file `barrier.sgc` and put it in the configured `--profile-dir`, it will be picked up automatically. Alternatively you can set a custom path to configuration file with argument `--config`.

From what I can see, nothing will be written to outside of the configured profile-dir!

If you want a minimal, copy-install, command-line-only deploy, you only need the following files, which you can copy from an existing installation, or extract from .exe installer using [innounp](http://innounp.sourceforge.net/):
```
barrierc.exe
barriers.exe
```

Due to dependency to OpenSSL, used for encryption (argument `--enable-crypto`), you need these files in addition (originally thought these were not needed unless using `--enable-crypto`, but found out they are a "hard" dependency):

```
libeay32.dll
ssleay32.dll
```

Optionally, if you want to be able to generate certificate, you also need these (on the server):
```
openssl.exe
barrier.conf
```

You will need the [Microsoft Visual C++ Redistributable for Visual Studio 2019](https://visualstudio.microsoft.com/downloads/) installed, although it *is* possible to copy-install `msvcp140.dll`, `vcruntime140.dll` and `vcruntime140_1.dll` into the application directory.

If you want to use encryption, you can generate certificate and fingerprint in a SSL directory structure the same way as the GUI application does, like this:
```
mkdir "D:\Applications\Barrier\SSL\Fingerprints" >NUL 2>&1
set OPENSSL_CONF=D:\Applications\Barrier\barrier.conf
"D:\Applications\Barrier\openssl.exe" req -x509 -nodes -days 365 -subj /CN=Barrier -newkey rsa:2048 -keyout "D:\Applications\Barrier\SSL\Barrier.pem" -out "D:\Applications\Barrier\SSL\Barrier.pem"
"D:\Applications\Barrier\openssl.exe" x509 -fingerprint -sha1 -noout -in "D:\Applications\Barrier\SSL\Barrier.pem" > "D:\Applications\Barrier\SSL\Fingerprints\Local.txt"
```

Edit: I realize `--profile-dir` is undocumented. Not listed in `--help` output, nor mentioned in [wiki](https://github.com/debauchee/barrier/wiki/Command-Line). That should be a small first step, I think.

---

> Commented on: 2020-12-05 by @Metroid-Prime

@albertony  thank you! I discovered this just yesterday and appreciate the summary.

I want to ask if the dev can make a "-portable" command argument that would do all this automatically. That would be wonderful.

---

> Commented on: 2020-12-05 by @Metroid-Prime

@albertony   unfortunately neither command works and it's a triple whammy.

This is what I used:

D:\Applications\Barrier\barrier.exe --profile-dir "D:\Applications\Barrier\settings\"

I got a folder created in:
C:\Users\Nightcrawler\AppData\Local\Barrier
C:\Users\Nightcrawler\AppData\Local\Barrier\SSL\
C:\Users\Nightcrawler\AppData\Local\Barrier\SSL\Fingerprints

and nothing I do stops Barrier from creating registry entries:
HKEY_CURRENT_USER\Software\Debauchee\Barrier
HKEY_LOCAL_MACHINE\SOFTWARE\Barrier

---

> Commented on: 2020-12-05 by @albertony

> D:\Applications\Barrier\barrier.exe --profile-dir "D:\Applications\Barrier\settings"

You are probably running the full GUI application? I haven't tried that myself, as I wanted to use it as a portable application from the start. Running the command line applications only, I don't get any of those folders or registry entries. But ok, if so, you have confirmed my suspicion that it does not work with that one.

Then, unfortunately, you will have to use the command line versions: barrier**s**.exe (server) and barrier**c**.exe (client), and create a text file with configuration for the server as described here https://github.com/debauchee/barrier/wiki/Command-Line#server-configuration-file. If you want to try this and have more questions, just ask.

**Comment by shymega** [comment]
Moving to discussion, this is a feature request.

**Comment by mqus** [comment]
For everyone interested in this: for running the client, you really only need input-leapc.exe (just install it on another PC first) and some SSL files in `%localappdata%\InputLeap\SSL` (can also be created on another PC, see above, this is the same as for barrier). You can then start it with `path-to\input-leapc.exe --name clientname serverhost`. This also does not need admin rights (another reason I don't like the current method of windows distribution via installer)

**Reply by spupuz** [reply]
> For everyone interested in this: for running the client, you really only need input-leapc.exe (just install it on another PC first) and some SSL files in `%localappdata%\InputLeap\SSL` (can also be created on another PC, see above, this is the same as for barrier). You can then start it with `path-to\input-leapc.exe --name clientname serverhost`. This also does not need admin rights (another reason I don't like the current method of windows distribution via installer)

can you explain better please

**Reply by matheusvellone** [reply]
I made it work here on my PC. Here are the steps

### 1. Copy the exe file

I've coppied the entire `InputLeap` folder from where I had admin access to install it to the other PC, but maybe just `input-leapc.exe` may be enough

### 2. Then the SSL Part
The `SSL` part is just a matter of adding the SHA256 on the server and client.
The folder is located at `C:\Users\USER\AppData\Local\InputLeap\SSL\Fingerprints`

Just to be clear (it always tricks me)
- Server = the computer with keyboard + mouse
- Client = computer to be controlled by server's keyboard + mouse

You need to add `v2:sha256:SERVER_FINGERPRINT` into the client's `TrustedServer.txt` file
And `v2:sha256:CLIENT_FINGERPRINT` into the server's `TrustedClients.txt` file

### 3. Running the command
Then just run the command in the client `input-leapc.exe --name CLIENT_NAME SERVER_IP`, where:
- `CLIENT_NAME` is the name to be identified on server, and to put it on the "screen grid"
- `SERVER_IP` is, well, the IP of the server

Of course, the server must be running.

I was having some trouble initially, and used `input-leapc.exe --name CLIENT_NAME -d DEBUG SERVER_IP` to enable DEBUG logs so I could check where it was failing

**Reply by Aymkdn** [reply]
Thank you soooo much! It was really helpful!

Some clarifications for people who would like to do it:
- You need to start InputLeap on the server first to get the `Local.txt` file created in `C:\Users\USER\AppData\Local\InputLeap\SSL\Fingerprints`, which contains the fingerprint
- Once the folder of InputLeap is copied to the client, you need to open `input-leap.exe` on the **client** which will create the fingerprint into `C:\Users\USER\AppData\Local\InputLeap\SSL\Fingerprints\Local.txt` – you can then exit it
- On the **client**, you must create the file `C:\Users\USER\AppData\Local\InputLeap\SSL\Fingerprints\TrustedServers.txt` (there is a "s" at "TrustedServers") with the `v2:sha256:SERVER_FINGERPRINT` from the server's `Local.txt`
- On the **server**, you must create the file `C:\Users\USER\AppData\Local\InputLeap\SSL\Fingerprints\TrustedClients.txt` with `v2:sha256:CLIENT_FINGERPRINT`
- On the **server**, open InputLeap, then go to **Configure Server** and click in a free cell to open a window where you can enter the **client's name**
- On the **client**, when running `input-leapc.exe --name CLIENT_NAME -d DEBUG SERVER_IP`, make sure to use the same `CLIENT_NAME` you use in the server's configuration
